app/frontend/src/features/search/SearchPage.tsx

```diff
@@ -61,13 +52,25 @@ export default function SearchPage() {
     undefined
   );
 
+  //callbacks provided to the map need a ref, or they won't get the latest value of selectedResult
+  const selectedResultRef = useRef<number | undefined>(undefined);
+  useEffect(() => {
+    selectedResultRef.current = selectedResult;
+  }, [selectedResult]);
```

Comment:
This took me a while...

Reply:
This defeats the purpose of using state and we're essentially mixing imperative and declarative ways of maintaining the same thing - we should choose one side and stick to it. Is it because we aren't consistently updating state to cause re-renders when it should?

Reply:
As you found in the other comment - the state value gets 'frozen' when the map is created because that's when the callback is set. What a good way to clean this up?

Reply:
If relying on state, usually the way to do it would be to clean up and re-register the event handlers in a `useEffect` whenever the state updates to ensure the callback gets the latest value, but that doesn't feel that different to what you're doing and it's more code...

I'm thinking it might be better to switch full imperative and hold that state in a ref and update it ourselves in the handlers in places where you set state? Or would that cause problems for when there is a query? My thinking is the map itself is already imperative and Mapbox is clearly doing some state management of the map on their end and it doesn't look like it'd play well with React state.

Reply:
Only problem is, the normal state is used too, to show which card is selected. If it's only in a ref, it won't re-render.

Reply:
I see, because that part is a declarative React component... the joy of having to mix the two together ☹️ 

In that case, I think I'd prefer cleaning up and re-subscribing the event handler whenever state updates, since that's the kind of things `useEffect` is designed for and so fits closer to the React way of doing things than ["swimming against the tide"](https://overreacted.io/a-complete-guide-to-useeffect/#swimming-against-the-tide).

This also lets the state be the single source of truth in this component so we don't have to figure out in certain bits, we have to use a "different copy" of the state (when they should be the same thing).